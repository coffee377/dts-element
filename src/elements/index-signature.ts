import {any_type} from '../constants';
import {Element} from '../element';
import {emit_optional} from '../helpers/emit-optional';
import {Stack} from '../stack';
import {mixin_class} from '../utils/mixin-class';
import {JSDocProtocol} from './declaration';
import {Type} from './type';

export interface IIndexSignatureRequiredParameters {
  name: string;
}

export interface IIndexSignatureOptionalParameters {
  jsdoc: string;
  kind: 'number' | 'string';
  type: Type;
  optional: boolean;
}

/**
 * ```ts
 * {
 *   [index: number]: any; // <--
 * }
 * ```
 */
export class IndexSignature
    extends Element<IIndexSignatureRequiredParameters, IIndexSignatureOptionalParameters> implements JSDocProtocol {

  public get default_parameters(): IIndexSignatureOptionalParameters {
    return {
      jsdoc: '',
      kind: 'string',
      type: any_type,
      optional: false,
    };
  }

  public _emit_jsdoc: () => string;

  private _instance_of_index_signature: true;

  // istanbul ignore next
  // tslint:disable-next-line prefer-function-over-method
  public _emit(_stack: Stack): string { return '[PLACEHOLDER]'; }

  public _emit_raw(stack: Stack): string {
    const {name, kind, type} = this.parameters;
    const optional = emit_optional(this.parameters.optional);
    return `[${name}: ${kind}]${optional}: ${type.emit(stack)};`;
  }

}

mixin_class(IndexSignature, [JSDocProtocol]);
