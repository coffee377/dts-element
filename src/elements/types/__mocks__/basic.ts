export class BasicType {

  public parameters: any;

  constructor(parameters: any) {
    this.parameters = parameters;
  }

  public _emit(): string {
    return `[BasicType ${this.parameters.name}]`;
  }

}
