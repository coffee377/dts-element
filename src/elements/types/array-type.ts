import {emit_brackets} from '../../helpers/emit-brackets';
import {Stack} from '../../stack';
import {Type} from '../type';

export interface IArrayTypeRequiredParameters {
  owned: Type;
}

// tslint:disable-next-line no-empty-interface
export interface IArrayTypeOptionalParameters {}

/**
 * ```ts
 * type X = some_type[];
 * //       ^^^^^^^^^^^
 * ```
 */
export class ArrayType extends Type<IArrayTypeRequiredParameters, IArrayTypeOptionalParameters> {

  private _instance_of_array_type: true;

  public get default_parameters(): IArrayTypeOptionalParameters {
    return {};
  }

  public _emit(stack: Stack): string {
    const {owned} = this.parameters;
    return `${emit_brackets(owned.emit(stack), owned)}[]`;
  }

}
