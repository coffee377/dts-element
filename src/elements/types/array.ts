import {AnyElement} from '../../element';
import {AnyType, Type} from '../type';

export interface IArrayTypeRequiredParameters {
  owned: AnyType;
}

// tslint:disable-next-line no-empty-interface
export interface IArrayTypeOptionalParameters {}

export class ArrayType extends Type<IArrayTypeRequiredParameters, IArrayTypeOptionalParameters> {

  public get default_parameters(): IArrayTypeOptionalParameters {
    return {};
  }

  public _emit(container: AnyElement | null): string {
    return `${this.parameters.owned._emit(container)}[]`;
  }

}
