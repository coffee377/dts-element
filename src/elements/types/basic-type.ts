import {Stack} from '../../stack';
import {Type} from '../type';

export interface IBasicTypeRequiredParameters {
  name: string;
}

// tslint:disable-next-line no-empty-interface
export interface IBasicTypeOptionalParameters {}

export class BasicType extends Type<IBasicTypeRequiredParameters, IBasicTypeOptionalParameters> {

  public get default_parameters(): IBasicTypeOptionalParameters {
    return {};
  }

  public _emit(_stack: Stack): string {
    return this.parameters.name;
  }

}
