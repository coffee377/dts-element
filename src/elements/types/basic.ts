import {AnyElement} from '../../element';
import {Type} from '../type';

export interface IBasicTypeRequiredParameters {
  name: string;
}

// tslint:disable-next-line no-empty-interface
export interface IBasicTypeOptionalParameters {}

export class BasicType extends Type<IBasicTypeRequiredParameters, IBasicTypeOptionalParameters> {

  public get default_parameters(): IBasicTypeOptionalParameters {
    return {};
  }

  public _emit(_container: AnyElement | null): string {
    return this.parameters.name;
  }

}
