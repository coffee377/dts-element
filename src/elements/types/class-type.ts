import {emit_declaration_type_name} from '../../helpers/emit-declaration-type-name';
import {emit_generics} from '../../helpers/emit-generics';
import {Stack} from '../../stack';
import {ClassDeclaration} from '../declarations/class-declaration';
import {Type} from '../type';

export interface IClassTypeRequiredParameters {
  owned: string | ClassDeclaration;
}

export interface IClassTypeOptionalParameters {
  generics: Type[];
}

export class ClassType extends Type<IClassTypeRequiredParameters, IClassTypeOptionalParameters> {

  public get default_parameters(): IClassTypeOptionalParameters {
    return {
      generics: [],
    };
  }

  public _emit(stack: Stack): string {
    const name = emit_declaration_type_name(this.parameters.owned);
    return `${name}${emit_generics(this.parameters.generics, stack)}`;
  }

}
