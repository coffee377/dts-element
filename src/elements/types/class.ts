import {Container} from '../../collections';
import {emit_generics} from '../../helpers/emit-generics';
import {Type} from '../type';

// TODO: support for ClassDeclaration
export interface IClassTypeRequiredParameters {
  class: string;
}

export interface IClassTypeOptionalParameters {
  generics: Type[];
}

export class ClassType extends Type<IClassTypeRequiredParameters, IClassTypeOptionalParameters> {

  public get default_parameters(): IClassTypeOptionalParameters {
    return {
      generics: [],
    };
  }

  public _emit(_container: Container): string {
    const name = this.parameters.class;
    return `${name}${emit_generics(this.parameters.generics, this)}`;
  }

}
