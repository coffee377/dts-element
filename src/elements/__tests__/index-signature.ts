jest.unmock('../index-signature');

import {IndexSignature} from '../index-signature';
import {BasicType} from '../types/basic';

describe('#emit()', () => {
  it('should return correctly', () => {
    expect(new IndexSignature({
      name: 'key',
      type: new BasicType({name: 'X'}),
    }).emit()).toMatchSnapshot();
  });
});
