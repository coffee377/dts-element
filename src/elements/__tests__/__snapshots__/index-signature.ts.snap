// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#emit() should return correctly 1`] = `"[key: string][emit-optional]: [BasicType X];"`;
